File: .github/scripts/check_js_syntax.py
#!/usr/bin/env python3
"""
check_js_syntax.py — 改善文書a 3.1
Extracts inline JS blocks from index.html and checks each for syntax errors
using `node --check`. Runs as a standalone script to avoid YAML/Python
quoting conflicts entirely.

Design decision (root fix):
  This project is Boring Technology (Vanilla JS, no TypeScript).
  Running tsc strict-mode on concatenated inline HTML scripts causes:
    1. Third-party minified blocks (KARTE etc.) trigger false errors
    2. Valid DOM patterns (Node.hasAttribute &&, EventTarget.closest) are
       flagged due to tsc type-narrowing — these are NOT real bugs
    3. MutationObserver NodeList typing is a tsc-only issue, not a runtime bug
  Solution: syntax-only check via `node --check` catches real breakage
  (SyntaxError, unexpected token etc.) without false positives.
  Type-level bug detection is fully delegated to ESLint (Step 24).
"""

import re
import sys
import subprocess
import tempfile
import os


def extract_js_blocks(html: str) -> list[str]:
    """Extract inline JS blocks, excluding third-party minified and non-JS types."""
    raw_pairs = re.findall(r"<script([^>]*)>(.*?)</script>", html, re.DOTALL | re.IGNORECASE)
    js_blocks = []
    for attrs, body in raw_pairs:
        a = attrs.lower()

        # Exclude external scripts
        if "src=" in a:
            continue

        # Exclude third-party minified single-line blocks (e.g. KARTE: !function(...){...})
        stripped = body.strip()
        if stripped.startswith("!function") and "\n" not in stripped:
            print(f"Skipped minified third-party block: {stripped[:60]}...")
            continue

        # Exclude non-JS types (application/ld+json, importmap, etc.)
        m = re.search(r'type\s*=\s*["\x27]([^"\x27]*)["\x27]', a, re.IGNORECASE)
        if m:
            t = m.group(1).strip()
            if t and t not in ("text/javascript", "module"):
                continue

        js_blocks.append(body)

    return js_blocks


def check_block(block: str, index: int) -> list[str]:
    """Run node --check on a single JS block. Returns list of error lines."""
    with tempfile.NamedTemporaryFile(
        mode="w", suffix=".js", delete=False, encoding="utf-8"
    ) as tf:
        tf.write(block)
        tf_path = tf.name

    try:
        result = subprocess.run(
            ["node", "--check", tf_path],
            capture_output=True,
            text=True,
        )
        if result.returncode != 0:
            # Replace temp path with readable label
            err = result.stderr.replace(tf_path, f"<inline-block-{index+1}>")
            return [err.strip()]
        return []
    finally:
        os.unlink(tf_path)


def main() -> int:
    html_path = "index.html"
    if not os.path.exists(html_path):
        print(f"::error::File not found: {html_path}")
        return 1

    with open(html_path, "r", encoding="utf-8") as f:
        html = f.read()

    blocks = extract_js_blocks(html)
    print(f"Checking {len(blocks)} JS blocks with node --check...")

    all_errors = []
    for i, block in enumerate(blocks):
        errs = check_block(block, i)
        if errs:
            for e in errs:
                print(f"::error::BLOCKING: Syntax error in inline JS block {i+1}:\n{e}")
            all_errors.extend(errs)

    if all_errors:
        print(f"::error::BLOCKING: {len(all_errors)} JS block(s) have syntax errors. Fix before merging.")
        return 1

    print(f"node --check: ALL {len(blocks)} blocks PASS")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: .github/scripts/check_js_syntax.mjs
#!/usr/bin/env node
/**
 * check_js_syntax.mjs — `npm run lint:js` の実体 (BLOCKING syntax gate)
 *
 * 【なぜこの file があるか】
 * 従来の `lint:js` は `node --check <file>` を 40 個並べた && chain だった。これは
 * **shipped JS 40 file 中 35 file (87.5%) に対して何も検査していなかった**。
 *
 *   実測 (node v26.3.0 / 2026-08-22):
 *     js/brand.js の createBrand 本体に `let let = 1;` を植える
 *       → `node --check js/brand.js`  … rc=0   ← 素通り
 *       → `npm run lint` (ESLint)     … rc=1   ← "Parsing error: The keyword 'let' is reserved"
 *
 * 原因は package.json に `"type": "module"` が無いこと。node は拡張子 .js を
 * CommonJS として parse し、失敗すると module-syntax detection で ESM として
 * 再 parse するが、**その再 parse の SyntaxError を報告せず exit 0 する**。
 * 同じ壊れた file を `"type":"module"` 下に置くと rc=1 になることも実測済み。
 *
 * package.json に `"type":"module"` を足す一行修正は採れない —— **e2e spec 62 file が
 * すべて `require()` を使っており**、BLOCKING merge gate である behavior e2e が丸ごと
 * 動かなくなる (実測)。よって「file ごとに、その file が実際に走る mode で parse する」
 * runner を置く。
 *
 * 【mode の決め方】
 * ESM 構文 (top-level の import / export) を持つ file は module mode、持たない file は
 * classic script mode。後者は index.html が `<script src>` (type=module なし) で読む
 * root script と service worker で、ブラウザでの読まれ方と一致する。
 * 判定は **コメントを除去してから** 行う (block comment 中の "export" を実装と誤認する
 * class を避ける —— この file を書く過程で実際に踏んだ)。
 *
 * 【非 goal (正直に記録)】
 * classic script が誤って `export` を獲得した場合、この gate は module mode へ分類して
 * 素通しする。ブラウザは classic script として throw するので実害はあるが、それは
 * 「syntax が妥当か」ではなく「module 配線が妥当か」の invariant で、Check 43d / 47 系の
 * 担当面。ここで曖昧に兼務させない。
 */

import { readFileSync } from 'node:fs';
import { spawnSync } from 'node:child_process';
import vm from 'node:vm';

const files = process.argv.slice(2);
if (files.length === 0) {
    console.error('check_js_syntax: no files given (usage: node check_js_syntax.mjs <file>...)');
    process.exit(2);
}

/** block/line コメントを潰す (文字列リテラル内の "//" を守るため、素朴だが保守的な走査)。 */
function stripComments(src) {
    let out = '';
    let i = 0;
    let quote = null;
    while (i < src.length) {
        const c = src[i];
        const next = src[i + 1];
        if (quote) {
            if (c === '\\') { out += '  '; i += 2; continue; }
            if (c === quote) { quote = null; }
            out += c === '\n' ? '\n' : ' ';
            i += 1;
            continue;
        }
        if (c === '"' || c === "'" || c === '`') { quote = c; out += ' '; i += 1; continue; }
        if (c === '/' && next === '*') {
            i += 2;
            while (i < src.length && !(src[i] === '*' && src[i + 1] === '/')) {
                out += src[i] === '\n' ? '\n' : ' ';
                i += 1;
            }
            i += 2;
            out += '  ';
            continue;
        }
        if (c === '/' && next === '/') {
            while (i < src.length && src[i] !== '\n') { out += ' '; i += 1; }
            continue;
        }
        out += c;
        i += 1;
    }
    return out;
}

const ESM_SYNTAX = /^[ \t]*(?:import|export)[\s{('"*;]/m;

let failed = 0;
for (const file of files) {
    let src;
    try {
        src = readFileSync(file, 'utf8');
    } catch (err) {
        console.error(`check_js_syntax: cannot read ${file} — ${err.message}`);
        failed += 1;
        continue;
    }

    const isModule = ESM_SYNTAX.test(stripComments(src));

    if (isModule) {
        // module mode は file 引数では正しく効かない (上の WHY) ので stdin 経由で渡す。
        // node は診断に "[stdin]" と出すため、file 名はこちらで補って帰属可能にする。
        const res = spawnSync(process.execPath, ['--input-type=module', '--check'], {
            input: src,
            encoding: 'utf8'
        });
        if (res.status !== 0) {
            const detail = (res.stderr || '').replace(/\[stdin\]/g, file).trimEnd();
            console.error(`✖ ${file} (module)\n${detail}\n`);
            failed += 1;
        }
    } else {
        // classic script は vm.Script が V8 の parser をそのまま通す (実行はしない)。
        try {
            new vm.Script(src, { filename: file });
        } catch (err) {
            console.error(`✖ ${file} (script)\n${err.stack || err.message}\n`);
            failed += 1;
        }
    }
}

if (failed > 0) {
    console.error(`check_js_syntax: ${failed} of ${files.length} file(s) failed to parse`);
    process.exit(1);
}
console.log(`check_js_syntax: ${files.length} files OK`);
